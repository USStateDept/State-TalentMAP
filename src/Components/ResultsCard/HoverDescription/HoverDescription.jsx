import { Component } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { throttle } from 'lodash';
import { Spring } from 'react-spring';
import FA from 'react-fontawesome';
import Linkify from 'react-linkify';
import InteractiveElement from '../../InteractiveElement';

class HoverDescription extends Component {
  constructor(props) {
    super(props);

    // Debouncing function to 260ms and binding this.
    // Used to create a delay and prevent multiple actions within a small
    // time frame from firing.
    this.toggleVisibility = throttle(this.toggleVisibility.bind(this),
      260, { leading: false, trailing: true });

    this.state = {
      expanded: false,
      cardHovered: false,
    };
  }

  toggleVisibility(shouldExpand = false) {
    this.setState({ expanded: !!shouldExpand });
  }

  toggle = () => {
    this.toggleVisibility(!this.state.expanded);
  };

  expand = () => {
    this.toggleVisibility(true);
  };

  close = () => {
    this.toggleVisibility(false);
  };

  toggleCardHovered(cardHovered = false) {
    this.setState({ cardHovered });
  }

  render() {
    const { isProjectedVacancy } = this.context;
    const { getOffsetPx, text, id } = this.props;
    const { expanded, cardHovered } = this.state;

    const getHeight = () => {
      if (expanded) {
        return getOffsetPx();
      }
      return cardHovered ? '31px' : '29px';
    };

    const height$ = getHeight();

    return (
      <Spring to={{ height: height$ }} config={{ friction: 21, tension: 175 }} >
        {
          ({ height }) =>
            (
              <div className="hover-description">
                <InteractiveElement
                  className="hover-button"
                  style={{ height }}
                  onMouseOver={this.expand}
                  onMouseLeave={this.close}
                  onClick={this.toggle}
                >
                  { !this.state.expanded && <FA name="angle-double-up" /> }
                  <Spring from={{ opacity: expanded ? 1 : 0 }}>
                    {
                      ({ opacity }) => (
                        <p style={{ opacity }}>
                          { expanded &&
                            <Linkify properties={{ target: '_blank' }}>
                              {text}
                              <Link
                                className="position-link"
                                to={`/${isProjectedVacancy ? 'vacancy' : 'details'}/${id}`}
                              >
                                View position
                              </Link>
                            </Linkify>
                          }
                        </p>
                      )
                    }
                  </Spring>
                </InteractiveElement>
              </div>
            )
        }
      </Spring>
    );
  }
}

HoverDescription.contextTypes = {
  isProjectedVacancy: PropTypes.bool,
};

HoverDescription.propTypes = {
  getOffsetPx: PropTypes.func.isRequired,
  text: PropTypes.string,
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
};

HoverDescription.defaultProps = {
  text: '',
};

export default HoverDescription;
