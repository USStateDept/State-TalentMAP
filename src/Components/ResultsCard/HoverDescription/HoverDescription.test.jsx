import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import HoverDescription from './HoverDescription';

describe('HoverDescriptionComponent', () => {
  const timeout = 500;
  const props = {
    getOffsetPx: () => '5px',
    text: 'Capsule description',
    id: 1,
  };

  it('is defined', () => {
    const wrapper = shallow(
      <HoverDescription
        {...props}
      />);
    expect(wrapper).toBeDefined();
  });

  it('sets state on toggleVisibility()', (done) => {
    const wrapper = shallow(
      <HoverDescription
        {...props}
      />);
    const instance = wrapper.instance();
    instance.toggleVisibility(true);
    setTimeout(() => {
      expect(instance.state.expanded).toBe(true);
      done();
    }, timeout);
  });

  it('sets state on toggle()', (done) => {
    const wrapper = shallow(
      <HoverDescription
        {...props}
      />);
    const instance = wrapper.instance();
    instance.toggle();
    setTimeout(() => {
      expect(instance.state.expanded).toBe(true);
      done();
    }, timeout);
  });

  it('sets state to true on expand()', (done) => {
    const wrapper = shallow(
      <HoverDescription
        {...props}
      />);
    const instance = wrapper.instance();
    instance.expand();
    setTimeout(() => {
      expect(instance.state.expanded).toBe(true);
      done();
    }, timeout);
  });

  it('sets state to close on close()', (done) => {
    const wrapper = shallow(
      <HoverDescription
        {...props}
      />);
    const instance = wrapper.instance();
    instance.close();
    setTimeout(() => {
      expect(instance.state.expanded).toBe(false);
      done();
    }, timeout);
  });

  it('sets state on toggle()', (done) => {
    const wrapper = shallow(
      <HoverDescription
        {...props}
      />);
    const instance = wrapper.instance();
    instance.toggle();
    setTimeout(() => {
      expect(instance.state.expanded).toBe(true);
      done();
    }, timeout);
  });

  it('sets state on toggleCardHovered()', (done) => {
    const wrapper = shallow(
      <HoverDescription
        {...props}
      />);
    const instance = wrapper.instance();
    instance.toggleCardHovered(true);
    setTimeout(() => {
      expect(instance.state.cardHovered).toBe(true);
      done();
    }, timeout);
  });

  it('matches snapshot', () => {
    const wrapper = shallow(
      <HoverDescription
        {...props}
      />);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it('matches snapshot when expanded === true', () => {
    const wrapper = shallow(
      <HoverDescription
        {...props}
      />);
    wrapper.instance().setState({ expanded: true });
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
