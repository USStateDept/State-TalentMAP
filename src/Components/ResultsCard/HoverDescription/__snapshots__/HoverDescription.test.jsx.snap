// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HoverDescriptionComponent matches snapshot 1`] = `
<e
  config={
    Object {
      "friction": 21,
      "tension": 175,
    }
  }
  from={Object {}}
  immediate={false}
  native={false}
  reset={false}
  to={
    Object {
      "height": "29px",
    }
  }
>
  <Component />
</e>
`;

exports[`HoverDescriptionComponent matches snapshot when expanded === true 1`] = `
<e
  config={
    Object {
      "friction": 21,
      "tension": 175,
    }
  }
  from={Object {}}
  immediate={false}
  native={false}
  reset={false}
  to={
    Object {
      "height": "5px",
    }
  }
>
  <Component />
</e>
`;
